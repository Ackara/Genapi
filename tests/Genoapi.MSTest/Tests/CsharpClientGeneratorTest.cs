using ApprovalTests;
using Microsoft.Build.Framework;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Shouldly;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using Tekcari.Gapi.Generators;
using Tekcari.Gapi.Generators.Csharp;
using Tekcari.Gapi.Serialization;
using Telerik.JustMock;
using Telerik.JustMock.Helpers;

namespace Tekcari.Gapi.Tests
{
	[TestClass]
	public class CsharpClientGeneratorTest
	{
		[TestMethod]
		[DynamicData(nameof(GetClientGeneratorCases), DynamicDataSourceType.Method)]
		public void Can_generate_csharp_api_client_support_classes(string documentPath, string serviceUrl)
		{
			var settings = new CsharpClientGeneratorSettings
			{
				BaseUrl = serviceUrl
			};
			RunGeneratorTest(new CsharpClientGenerator(settings), documentPath, (x) => x.Tag == "native");
		}

		[TestMethod]
		[DynamicData(nameof(GetClientGeneratorCases), DynamicDataSourceType.Method)]
		public void Can_generate_csharp_api_client_components(string documentPath, string serviceUrl)
		{
			var settings = new CsharpClientGeneratorSettings
			{
				BaseUrl = serviceUrl
			};
			RunGeneratorTest(new CsharpClientGenerator(settings), documentPath, (x) => x.Tag == "component");
		}

		[TestMethod]
		[DynamicData(nameof(GetClientGeneratorCases), DynamicDataSourceType.Method)]
		public void Can_generate_csharp_api_client_endpoints(string documentPath, string serviceUrl)
		{
			var settings = new CsharpClientGeneratorSettings
			{
				BaseUrl = serviceUrl
			};
			RunGeneratorTest(new CsharpClientGenerator(settings), documentPath, (x) => x.Tag == "client");
		}

		[TestMethod]
		[DynamicData(nameof(GetClientGeneratorCases), DynamicDataSourceType.Method)]
		public void Can_generate_csharp_api_client_test_suite(string documentPath, string serviceUrl)
		{
			var settings = new CsharpClientTestSuiteGeneratorSettings
			{
				ServiceUrl = serviceUrl
			};
			RunGeneratorTest(new CsharpClientTestSuiteGenerator(settings), documentPath);
		}

		[TestMethod]
		public void Can_invoke_generate_csharp_api_client_task()
		{
			// Arrange

			var specificationFile = Mock.Create<ITaskItem>();
			specificationFile.Arrange(x => x.GetMetadata(Arg.Is("FullPath")))
				.Returns(TestData.GetFilePath("petstore.json"));

			string resultFilePath = Path.Combine(Path.GetTempPath(), "GeneratedClient.txt");
			var outFile = Mock.Create<ITaskItem>();
			outFile.Arrange(x => x.GetMetadata(Arg.Is("FullPath")))
				.Returns(resultFilePath);

			var task = new Targets.GenerateCsharpApiClient
			{
				SourceFile = specificationFile,
				DestinationFile = outFile
			};

			// Act

			var sucess = task.Execute();

			// Assert

			sucess.ShouldBeTrue();
			Approvals.VerifyFile(resultFilePath);
		}

		[TestMethod]
		public void Can_invoke_generate_csharp_api_client_test_suite_task()
		{
			// Arrange

			var specificationFile = Mock.Create<ITaskItem>();
			specificationFile.Arrange(x => x.GetMetadata(Arg.Is("FullPath")))
				.Returns(TestData.GetFilePath("petstore.json"));

			string resultFilePath = Path.Combine(Path.GetTempPath(), "GeneratedClientTest.txt");
			var outFile = Mock.Create<ITaskItem>();
			outFile.Arrange(x => x.GetMetadata(Arg.Is("FullPath")))
				.Returns(resultFilePath);

			var task = new Targets.GenerateCsharpApiClientTestSuite
			{
				ServiceUrl = "https://petstore3.swagger.io/api/v3",
				SourceFile = specificationFile,
				DestinationFile = outFile,
			};

			// Act

			var sucess = task.Execute();

			// Assert

			sucess.ShouldBeTrue();
			Approvals.VerifyFile(resultFilePath);
		}

		[TestMethod]
		public void Can_extract_using_statments()
		{
			// Arrange

			string sample = @"
using System;
using System.Linq;
using System.Data;

namespace foo
{
	publice class ClassA {}
	publice class ClassA {}
}

using System;
using System.Linq;
using System.Data;

namespace bar
{
	publice class ClassC {}
	publice class ClassD {}
}
";

			// Act

			CsharpGenerator.ExtractUsingStatements(sample, out string code, out string[] usings);

			// Assert
			code.ShouldNotMatch(@"(?i)using\s+[A-Z0-9_\.]+;");
			usings.ShouldNotBeEmpty();
			usings.Length.ShouldBe(3);
		}

		internal static void RunGeneratorTest(IGenerator generator, string documentPath, Func<FileResult, bool> filter = default)
		{
			var scenario = ApprovalTests.Namers.ApprovalResults.ForScenario(Path.GetFileName(documentPath));
			if (filter == default) filter = (x) => !string.IsNullOrEmpty(x.Name);

			var spec = DocumentLoader.Load(documentPath);
			var fileList = generator.Generate(spec);
			var generatedCode = MergeAndAnalyze(fileList.Where(filter));

			fileList.ShouldNotBeEmpty();
			Approvals.Verify(generatedCode);
		}

		#region Backing Members

		public static IEnumerable<object[]> GetClientGeneratorCases()
		{
			yield return new object[] { TestData.GetFilePath("openapi.json"), "" };
			//yield return new object[] { TestData.GetFilePath("petstore.json"), "https://petstore3.swagger.io/api/v3" };
			//yield return new object[] { TestData.GetFilePath("plaid.yml") };
		}

		internal static string MergeAndAnalyze(IEnumerable<FileResult> fileList)
		{
			var builder = new StringBuilder();
			foreach (var file in fileList)
			{
				builder.Append("== ");
				builder.AppendLine(file.Name);
				builder.AppendLine(string.Concat(Enumerable.Repeat('=', 50)));
				builder.AppendLine();

				AppendErrors(builder, file.Content);
				builder.AppendLine(file.Content);
				builder.AppendLine().AppendLine();
			}
			return builder.ToString();
		}

		private static void AppendErrors(StringBuilder builder, string csharp, params string[] ignore)
		{
			if (string.IsNullOrEmpty(csharp)) return;

			var snippet = Microsoft.CodeAnalysis.CSharp.CSharpSyntaxTree.ParseText(csharp);
			var errors = from x in snippet.GetDiagnostics()
						 where
							x.Severity >= Microsoft.CodeAnalysis.DiagnosticSeverity.Warning
							&&
							ignore.Contains(x.Id) == false
						 select $"[{x.Id}] Line {x.Location.GetMappedLineSpan().StartLinePosition.Line + 1}: {x.GetMessage()}";

			if (errors.Any())
			{
				builder
					.Append(string.Concat(Enumerable.Repeat('*', 21)))
					.Append(" ERRORS ")
					.AppendLine(string.Concat(Enumerable.Repeat('*', 21)))
					.AppendLine()
					.AppendJoin("\r\n", errors)
					.AppendLine()
					.AppendLine()
					.AppendLine(string.Concat(Enumerable.Repeat('*', 50)))
					.AppendLine()
					.AppendLine();
			}
		}

		#endregion Backing Members
	}
}