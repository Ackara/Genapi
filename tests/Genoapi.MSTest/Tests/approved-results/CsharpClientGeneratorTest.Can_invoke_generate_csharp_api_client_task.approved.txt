using System.Text.Json.Serialization;
using System.Threading.Tasks;
using System.Text.Json;
using System.Net.Http;
using System.Linq;
using System.Collections.Generic;
using System.IO;
using System;
using Microsoft.Extensions.DependencyInjection;

namespace Tekcari.Gapi
{
	public partial class GeneratedClient
	{
		public GeneratedClient(string baseUrl, JsonSerializerOptions jsonSerializerOptions = default, IHttpClientFactory httpClientFactory = default)
		{
			_baseUrl = baseUrl?.Trim('/');
			_httpClientFactory = httpClientFactory ?? GetHttpClientFactory() ?? throw new ArgumentNullException(nameof(httpClientFactory));
			_serializerOptions = jsonSerializerOptions ?? new JsonSerializerOptions
			{
				PropertyNameCaseInsensitive = true,
				PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
				DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
			};
			Initialize();
		}
		
		/// <summary>Update an existing pet.</summary>
		/// <remarks>Update an existing pet by Id.</remarks>
		/// <param name="pet">Update an existent pet in the store.</param>
		public Task<Response<Pet>> UpdatePetAsync(Pet pet)
		{
			var request = new HttpRequestMessage(new HttpMethod("PUT"), Url($"/pet"));
			request.Content = ToJson(pet);
			return SendRequestAsync<Pet>(request);
		}

		/// <summary>Add a new pet to the store.</summary>
		/// <remarks>Add a new pet to the store.</remarks>
		/// <param name="pet">Create a new pet in the store.</param>
		public Task<Response<Pet>> AddPetAsync(Pet pet)
		{
			var request = new HttpRequestMessage(new HttpMethod("POST"), Url($"/pet"));
			request.Content = ToJson(pet);
			return SendRequestAsync<Pet>(request);
		}

		/// <summary>Finds Pets by status.</summary>
		/// <remarks>Multiple status values can be provided with comma separated strings.</remarks>
		/// <param name="status">Status values that need to be considered for filter.</param>
		public Task<Response<Pet[]>> FindPetsByStatusAsync(string status = default)
		{
			var request = new HttpRequestMessage(new HttpMethod("GET"), Url($"/pet/findByStatus?status={status}"));
			return SendRequestAsync<Pet[]>(request);
		}

		/// <summary>Finds Pets by tags.</summary>
		/// <remarks>Multiple tags can be provided with comma separated strings. Use tag1, tag2, tag3 for testing.</remarks>
		/// <param name="tags">Tags to filter by.</param>
		public Task<Response<Pet[]>> FindPetsByTagsAsync(params string[] tags)
		{
			var request = new HttpRequestMessage(new HttpMethod("GET"), Url($"/pet/findByTags?{GetQueryList("tags", tags)}"));
			return SendRequestAsync<Pet[]>(request);
		}

		/// <summary>Find pet by ID.</summary>
		/// <remarks>Returns a single pet.</remarks>
		/// <param name="petId">ID of pet to return.</param>
		public Task<Response<Pet>> GetPetByIdAsync(long petId)
		{
			var request = new HttpRequestMessage(new HttpMethod("GET"), Url($"/pet/{petId}"));
			return SendRequestAsync<Pet>(request);
		}

		/// <summary>Updates a pet in the store with form data.</summary>
		/// <param name="petId">ID of pet that needs to be updated.</param>
		/// <param name="name">Name of pet that needs to be updated.</param>
		/// <param name="status">Status of pet that needs to be updated.</param>
		public Task<Response> UpdatePetWithFormAsync(long petId, string name = default, string status = default)
		{
			var request = new HttpRequestMessage(new HttpMethod("POST"), Url($"/pet/{petId}?name={name}&status={status}"));
			return SendRequestAsync(request);
		}

		/// <summary>Deletes a pet.</summary>
		/// <param name="petId">Pet id to delete.</param>
		/// <param name="api_key"></param>
		public Task<Response> DeletePetAsync(long petId, string api_key = default)
		{
			var request = new HttpRequestMessage(new HttpMethod("DELETE"), Url($"/pet/{petId}"));
			request.Headers.Add("api_key", api_key);
			return SendRequestAsync(request);
		}

		/// <summary>uploads an image.</summary>
		/// <param name="data"></param>
		/// <param name="petId">ID of pet to update.</param>
		/// <param name="additionalMetadata">Additional Metadata.</param>
		public Task<Response<ApiResponse>> UploadFileAsync(byte[] data, long petId, string additionalMetadata = default)
		{
			var request = new HttpRequestMessage(new HttpMethod("POST"), Url($"/pet/{petId}/uploadImage?additionalMetadata={additionalMetadata}"));
			return SendRequestAsync<ApiResponse>(request);
		}

		/// <summary>Returns pet inventories by status.</summary>
		/// <remarks>Returns a map of status codes to quantities.</remarks>
		public Task<Response<string>> GetInventoryAsync()
		{
			var request = new HttpRequestMessage(new HttpMethod("GET"), Url($"/store/inventory"));
			return SendRequestAsync<string>(request);
		}

		/// <summary>Place an order for a pet.</summary>
		/// <remarks>Place a new order in the store.</remarks>
		/// <param name="order"></param>
		public Task<Response<Order>> PlaceOrderAsync(Order order)
		{
			var request = new HttpRequestMessage(new HttpMethod("POST"), Url($"/store/order"));
			request.Content = ToJson(order);
			return SendRequestAsync<Order>(request);
		}

		/// <summary>Find purchase order by ID.</summary>
		/// <remarks>For valid response try integer IDs with value <= 5 or > 10. Other values will generated exceptions.</remarks>
		/// <param name="orderId">ID of order that needs to be fetched.</param>
		public Task<Response<Order>> GetOrderByIdAsync(long orderId)
		{
			var request = new HttpRequestMessage(new HttpMethod("GET"), Url($"/store/order/{orderId}"));
			return SendRequestAsync<Order>(request);
		}

		/// <summary>Delete purchase order by ID.</summary>
		/// <remarks>For valid response try integer IDs with value < 1000. Anything above 1000 or nonintegers will generate API errors.</remarks>
		/// <param name="orderId">ID of the order that needs to be deleted.</param>
		public Task<Response> DeleteOrderAsync(long orderId)
		{
			var request = new HttpRequestMessage(new HttpMethod("DELETE"), Url($"/store/order/{orderId}"));
			return SendRequestAsync(request);
		}

		/// <summary>Create user.</summary>
		/// <remarks>This can only be done by the logged in user.</remarks>
		/// <param name="user">Created user object.</param>
		public Task<Response> CreateUserAsync(User user)
		{
			var request = new HttpRequestMessage(new HttpMethod("POST"), Url($"/user"));
			request.Content = ToJson(user);
			return SendRequestAsync(request);
		}

		/// <summary>Creates list of users with given input array.</summary>
		/// <remarks>Creates list of users with given input array.</remarks>
		/// <param name="user"></param>
		public Task<Response<User>> CreateUsersWithListInputAsync(User[] user)
		{
			var request = new HttpRequestMessage(new HttpMethod("POST"), Url($"/user/createWithList"));
			request.Content = ToJson(user);
			return SendRequestAsync<User>(request);
		}

		/// <summary>Logs user into the system.</summary>
		/// <param name="username">The user name for login.</param>
		/// <param name="password">The password for login in clear text.</param>
		public Task<Response<string>> LoginUserAsync(string username = default, string password = default)
		{
			var request = new HttpRequestMessage(new HttpMethod("GET"), Url($"/user/login?username={username}&password={password}"));
			return SendRequestAsync<string>(request);
		}

		/// <summary>Logs out current logged in user session.</summary>
		public Task<Response> LogoutUserAsync()
		{
			var request = new HttpRequestMessage(new HttpMethod("GET"), Url($"/user/logout"));
			return SendRequestAsync(request);
		}

		/// <summary>Get user by user name.</summary>
		/// <param name="username">The name that needs to be fetched. Use user1 for testing. .</param>
		public Task<Response<User>> GetUserByNameAsync(string username)
		{
			var request = new HttpRequestMessage(new HttpMethod("GET"), Url($"/user/{username}"));
			return SendRequestAsync<User>(request);
		}

		/// <summary>Update user.</summary>
		/// <remarks>This can only be done by the logged in user.</remarks>
		/// <param name="user">Update an existent user in the store.</param>
		/// <param name="username">name that need to be deleted.</param>
		public Task<Response> UpdateUserAsync(User user, string username)
		{
			var request = new HttpRequestMessage(new HttpMethod("PUT"), Url($"/user/{username}"));
			request.Content = ToJson(user);
			return SendRequestAsync(request);
		}

		/// <summary>Delete user.</summary>
		/// <remarks>This can only be done by the logged in user.</remarks>
		/// <param name="username">The name that needs to be deleted.</param>
		public Task<Response> DeleteUserAsync(string username)
		{
			var request = new HttpRequestMessage(new HttpMethod("DELETE"), Url($"/user/{username}"));
			return SendRequestAsync(request);
		}


		internal async Task<Response> SendRequestAsync(HttpRequestMessage request)
		{
#if DEBUG
			PrintToDebugWindow(request);
#endif
			HttpClient client = _httpClientFactory.CreateClient();
			using (HttpResponseMessage response = await client.SendAsync(request))
			{
#if DEBUG
				PrintToDebugWindow(response);
#endif
				if (response.IsSuccessStatusCode)
				{
					return new Response((int)response.StatusCode, response.ReasonPhrase);
				}
				else
				{
					string text = await response.Content.ReadAsStringAsync();
					if (text.Length == 0) return new Response((int)response.StatusCode, response.ReasonPhrase);
					else
					{
						return new Response((int)response.StatusCode, text);
					}
				}
			}
		}

		internal async Task<Response<T>> SendRequestAsync<T>(HttpRequestMessage request)
		{
#if DEBUG
			PrintToDebugWindow(request);
#endif
			HttpClient client = _httpClientFactory.CreateClient();
			using (HttpResponseMessage response = await client.SendAsync(request))
			{
#if DEBUG
				PrintToDebugWindow(response);
#endif
				string json = await response.Content.ReadAsStringAsync();
				if (string.IsNullOrEmpty(json)) return new Response<T>(default, (int)response.StatusCode, response.ReasonPhrase);

				if (response.IsSuccessStatusCode)
				{
					if (typeof(IConvertible).IsAssignableFrom(typeof(T)))
						return new Response<T>((T)Convert.ChangeType(json, typeof(T)), (int)response.StatusCode);
					else
						return new Response<T>(
							JsonSerializer.Deserialize<T>(json, _serializerOptions),
							(int)response.StatusCode,
							response.ReasonPhrase
						);
				}
				else
				{
					return new Response<T>(default, (int)response.StatusCode, json);
				}
			}
		}

		internal StringContent ToJson(object obj)
		{
			string json = System.Text.Json.JsonSerializer.Serialize(obj, _serializerOptions);
			return new StringContent(json, System.Text.Encoding.UTF8, "application/json");
		}

		partial void Initialize();

		#region Backing Members

		private readonly string _baseUrl;
		private readonly IHttpClientFactory _httpClientFactory;
		private JsonSerializerOptions _serializerOptions;

		private Uri Url(string path) => new Uri(string.Concat(_baseUrl, path), UriKind.RelativeOrAbsolute);
		private static IHttpClientFactory GetHttpClientFactory() => new ServiceCollection().AddHttpClient().BuildServiceProvider().GetService<IHttpClientFactory>();

		private static string GetQueryList(string name, object[] args) => string.Join("&", from x in args select $"{name}={args}");

		private void PrintToDebugWindow(HttpRequestMessage message)
		{
#if DEBUG
			System.Diagnostics.Debug.Write("===== HTTP Request ");
			System.Diagnostics.Debug.WriteLine(string.Concat(Enumerable.Repeat('=', 50)));
			System.Diagnostics.Debug.WriteLine($"{message.Method.Method}: {message.RequestUri}");
			System.Diagnostics.Debug.WriteLine(message.Content?.ReadAsStringAsync().Result);
			System.Diagnostics.Debug.WriteLine(string.Empty);
#endif
		}

		private void PrintToDebugWindow(HttpResponseMessage message)
		{
#if DEBUG
			System.Diagnostics.Debug.Write($"===== HTTP Response ({(int)message.StatusCode}) ");
			System.Diagnostics.Debug.WriteLine(string.Concat(Enumerable.Repeat('=', 50)));
			System.Diagnostics.Debug.WriteLine($"{message.RequestMessage.Method.Method}: {message.RequestMessage.RequestUri}");
			System.Diagnostics.Debug.WriteLine($"Status: {message.ReasonPhrase} ({(int)message.StatusCode})");
			System.Diagnostics.Debug.WriteLine(message.Content?.ReadAsStringAsync().Result);
			System.Diagnostics.Debug.WriteLine(string.Empty);
#endif
		}

		#endregion Backing Members
	}
}

namespace Tekcari.Gapi
{
	public partial class ApiResponse
	{
		[JsonPropertyName("code")]
		public int Code { get; set; }

		[JsonPropertyName("type")]
		public string Type { get; set; }

		[JsonPropertyName("message")]
		public string Message { get; set; }

	}
}

namespace Tekcari.Gapi
{
	public partial class Pet
	{
		[JsonPropertyName("id")]
		public long Id { get; set; }

		[JsonPropertyName("name")]
		public string Name { get; set; }

		[JsonPropertyName("category")]
		public Category Category { get; set; }

		[JsonPropertyName("photoUrls")]
		public string[] PhotoUrls { get; set; }

		[JsonPropertyName("tags")]
		public Tag[] Tags { get; set; }

		/// <summary>pet status in the store</summary>
		[JsonPropertyName("status")]
		public string Status { get; set; }

	}
}

namespace Tekcari.Gapi
{
	public partial class Tag
	{
		[JsonPropertyName("id")]
		public long Id { get; set; }

		[JsonPropertyName("name")]
		public string Name { get; set; }

	}
}

namespace Tekcari.Gapi
{
	public partial class User
	{
		[JsonPropertyName("id")]
		public long Id { get; set; }

		[JsonPropertyName("username")]
		public string Username { get; set; }

		[JsonPropertyName("firstName")]
		public string FirstName { get; set; }

		[JsonPropertyName("lastName")]
		public string LastName { get; set; }

		[JsonPropertyName("email")]
		public string Email { get; set; }

		[JsonPropertyName("password")]
		public string Password { get; set; }

		[JsonPropertyName("phone")]
		public string Phone { get; set; }

		/// <summary>User Status</summary>
		[JsonPropertyName("userStatus")]
		public int UserStatus { get; set; }

	}
}

namespace Tekcari.Gapi
{
	public partial class Category
	{
		[JsonPropertyName("id")]
		public long Id { get; set; }

		[JsonPropertyName("name")]
		public string Name { get; set; }

	}
}

namespace Tekcari.Gapi
{
	public partial class Address
	{
		[JsonPropertyName("street")]
		public string Street { get; set; }

		[JsonPropertyName("city")]
		public string City { get; set; }

		[JsonPropertyName("state")]
		public string State { get; set; }

		[JsonPropertyName("zip")]
		public string Zip { get; set; }

	}
}

namespace Tekcari.Gapi
{
	public partial class Customer
	{
		[JsonPropertyName("id")]
		public long Id { get; set; }

		[JsonPropertyName("username")]
		public string Username { get; set; }

		[JsonPropertyName("address")]
		public Address[] Address { get; set; }

	}
}

namespace Tekcari.Gapi
{
	public partial class Order
	{
		[JsonPropertyName("id")]
		public long Id { get; set; }

		[JsonPropertyName("petId")]
		public long PetId { get; set; }

		[JsonPropertyName("quantity")]
		public int Quantity { get; set; }

		[JsonPropertyName("shipDate")]
		public DateTime ShipDate { get; set; }

		/// <summary>Order Status</summary>
		[JsonPropertyName("status")]
		public string Status { get; set; }

		[JsonPropertyName("complete")]
		public bool Complete { get; set; }

	}
}

namespace Tekcari.Gapi
{
	[System.Diagnostics.DebuggerDisplay("{" + nameof(Format) + "(),nq}")]
	/// <summary>
	/// Represents response from API server.
	/// </summary>
	public readonly struct Response
	{
		/// <summary>
		/// Initializes a new instance of the <see cref="Response"/> struct.
		/// </summary>
		/// <param name="statusCode">The status code.</param>
		/// <param name="message">The message.</param>
		public Response(int statusCode, string message = default)
			: this(IsGood(statusCode), statusCode, message)
		{
		}

		/// <summary>
		/// Initializes a new instance of the <see cref="Response"/> struct.
		/// </summary>
		/// <param name="success">if set to <c>true</c> [success].</param>
		/// <param name="statusCode">The status code.</param>
		/// <param name="message">The message.</param>
		public Response(bool success, int statusCode, string message)
		{
			Message = message;
			Succeeded = success;
			StatusCode = statusCode;
		}

		/// <summary>
		/// Gets the HTTP status code.
		/// </summary>
		/// <value>The HTTP status code.</value>
		public int StatusCode { get; }

		/// <summary>
		/// Gets the error message.
		/// </summary>
		/// <value>The message.</value>
		public string Message { get; }

		/// <summary>
		/// Gets a value indicating whether the corresponding request is succeeded.
		/// </summary>
		/// <value><c>true</c> if succeeded; otherwise, <c>false</c>.</value>
		public bool Succeeded { get; }

		/// <summary>
		/// Gets a value indicating whether the corresponding request is failed.
		/// </summary>
		/// <value><c>true</c> if failed; otherwise, <c>false</c>.</value>
		public bool Failed
		{
			get => Succeeded == false;
		}

		/// <summary>
		/// Convert the value of this instance to a <see cref="System.String"/>.
		/// </summary>
		/// <returns>
		/// A <see cref="System.String" /> that represents this instance.
		/// </returns>
		public override string ToString() => Message;

		/// <summary>
		/// Returns a list of errors returned by the server.
		/// </summary>
		public IEnumerable<ResponseError> GetErrors() => ParseErrors(Message);

		internal static IEnumerable<ResponseError> ParseErrors(string message)
		{
			if (string.IsNullOrEmpty(message)) yield break;

			string[] errors = message.Split(new string[] { "\n", "\r\n" }, StringSplitOptions.RemoveEmptyEntries);
			foreach (string err in errors)
	        {
				if (err.Length > 5 && err[4] == ':')
				{
					int.TryParse(err.Substring(0, 4), out int code);
					yield return new ResponseError(err.Substring(5), code);
				}
				else yield return new ResponseError(err, 0);
		    }
		}

		internal string Format() => $"({StatusCode}): {Message}".Trim(' ', ':');

		internal static bool IsGood(int code) => code >= 200 && code < 300;

		#region operators

		public static implicit operator bool(Response x) => x.Succeeded;

		#endregion operators
	}

	[System.Diagnostics.DebuggerDisplay("{" + nameof(Format) + "(),nq}")]
	/// <summary>
	/// Represents response data from API server.
	/// </summary>
	/// <typeparam name="T"></typeparam>
	public readonly struct Response<T>
	{
		/// <summary>
		/// Initializes a new instance of the <see cref="Response{T}"/> struct.
		/// </summary>
		/// <param name="data">The data.</param>
		/// <param name="statusCode">The status code.</param>
		/// <param name="message">The message.</param>
		public Response(T data, int statusCode = 200, string message = default)
			: this(data, Response.IsGood(statusCode), statusCode, message) { }

		/// <summary>
		/// Initializes a new instance of the <see cref="Response{T}"/> struct.
		/// </summary>
		/// <param name="data">The data.</param>
		/// <param name="success">if set to <c>true</c> [success].</param>
		/// <param name="status">The status.</param>
		/// <param name="message">The message.</param>
		public Response(T data, bool success, int status, string message)
		{
			Data = data;
			Succeeded = success;
			StatusCode = status;
			Message = message;
		}

		/// <summary>
		/// Gets the data.
		/// </summary>
		/// <value>The data.</value>
		public T Data { get; }

		/// <summary>
		/// Gets the HTTP status code.
		/// </summary>
		/// <value>The HTTP status code.</value>
		public int StatusCode { get; }

		/// <summary>
		/// Gets the error message.
		/// </summary>
		/// <value>The message.</value>
		public string Message { get; }

		/// <summary>
		/// Gets a value indicating whether the corresponding request is succeeded.
		/// </summary>
		/// <value><c>true</c> if succeeded; otherwise, <c>false</c>.</value>
		public bool Succeeded { get; }

		/// <summary>
		/// Gets a value indicating whether the corresponding request is failed.
		/// </summary>
		/// <value><c>true</c> if failed; otherwise, <c>false</c>.</value>
		public bool Failed
		{
			get => Succeeded == false;
		}

		/// <summary>
		/// Convert the value of this instance to a <see cref="System.String"/>.
		/// </summary>
		/// <returns>
		/// A <see cref="System.String" /> that represents this instance.
		/// </returns>
		public override string ToString() => Message;

		/// <summary>
		/// Returns a list of errors returned by the server.
		/// </summary>
		public IEnumerable<ResponseError> GetErrors() => Response.ParseErrors(Message);

		internal string Format() => $"({StatusCode}): {Message}".Trim(' ', ':');

		#region operators

		public static implicit operator T(Response<T> x) => x.Data;

		public static explicit operator bool(Response<T> x) => x.Succeeded;

		public static implicit operator Response(Response<T> x) => new Response(x.Succeeded, x.StatusCode, x.Message);

		#endregion operators
	}

	/// <summary>
	/// Represents an error returned by a server.
	/// </summary>
	public readonly struct ResponseError
	{
		/// <summary>
		/// /// Initializes a new instance of the <see cref="ResponseError"/> struct.
		/// </summary>
		/// <param name="message">The error message.</param>
		/// <param name="code">The error code.</param>
		public ResponseError(string message, int code)
		{
			ErrorCode = code;
			Message = message;
		}

		/// <summary>
		/// The error code.
		/// </summary>
		public int ErrorCode { get; }

		/// <summary>
		/// The error message.
		/// </summary>
		public string Message { get; }
	}
}

